<a href="https://colab.research.google.com/github/opendatas2017/NMC/blob/main/NMCourse_prec_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Точность


Ссылки

https://toomanydigits.online/Block1/Sem1/4.html

https://docs-python.ru/tutorial/operatsii-chislami-python/problemy-chisel-plavajuschej-zapjatoj/

### Основная информация и особенности


In [ ]:
import numpy as np
import sys

print("Numpy version is ",np.__version__)  # Должно показать версию, например 2.0.2

print("Python version is ",sys.version)

Numpy version is  2.0.2
Python version is  3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]


Машинный эпсилон

### 🔬 Машинный эпсилон (ε): предел точности вычислений с плавающей запятой

**Что такое машинный эпсилон?**  
Минимальное число $\varepsilon > 0$, такое что $1.0 + \varepsilon > 1.0$ в арифметике с плавающей запятой.

**Почему это важно?**
- Определяет **предел относительной точности** вычислений (~15-16 знаков для float64)
- Объясняет, почему `0.1 + 0.2 != 0.3` в компьютере
- Задаёт нижнюю границу шага $h$ в численных методах (дифференцирование, интегрирование)

**Стандарт IEEE 754:**
| Тип | Размер | ε (машинный эпсилон) | Примерный диапазон |
|-----|--------|---------------------|-------------------|
| **float16** | 16 бит | ~$10^{-3}$ | $[6\cdot10^{-8}, 6\cdot10^{4}]$ |
| **float32** | 32 бит | ~$10^{-7}$ | $[10^{-38}, 10^{38}]$ |
| **float64** | 64 бит | ~$10^{-16}$ | $[10^{-308}, 10^{308}]$ |



In [ ]:
# === 0. Импорт библиотек ===
import sys
import numpy as np

# === 1. Эталонные значения ε из библиотек ===
print("📋 МАШИННЫЙ ЭПСИЛОН (ε) — минимальное число, для которого 1+ε > 1:\n")

# Стандартные значения для разных типов данных
print(f"{'Тип данных':<15} | {'Машинный ε':>15} | {'Значимых цифр':>15}")
print("-" * 52)

# Python (всегда float64)
eps_py = sys.float_info.epsilon

print(f"{'sys.float_info':<15} | {eps_py:15.2e} | {-np.log10(eps_py):>14.0f} знаков")

# NumPy: основные типы
for dtype, name in [(np.float64, 'float64'), (np.float32, 'float32'), (np.float16, 'float16')]:
    eps_np = np.finfo(dtype).eps
    digits = -np.log10(eps_np)
    # ИСПРАВЛЕНО: округление до целых через .0f
    print(f"{name:<15} | {eps_np:15.2e} | {digits:>14.0f} знаков")

# === 2. Вычисление ε "в лоб" — алгоритмическая демонстрация ===
print(f"\n🔍 Вычисляем ε для float64 алгоритмически:")
print("   Алгоритм: делим ε пополам, пока 1+ε не перестанет отличаться от 1")

eps_computed = 1.0
iteration = 0

# Цикл деления пополам: ищем границу, где арифметика "не видит" разницы
while (1.0 + eps_computed/2) > 1.0:
    eps_computed /= 2
    iteration += 1

print(f"   Итераций: {iteration}")
print(f"   🏎️  Вычисленный ε = {eps_computed:.2e}")

# === 3. Проверка: демонстрируем "границу видимости" ===
print(f"\n✅ Проверка результата:")
print(f"   1.0 + ε/2 = {1.0 + eps_computed/2:.17f}  → равно 1.0? {1.0 + eps_computed/2 == 1.0}")
print(f"   1.0 + ε   = {1.0 + eps_computed:.17f}  → равно 1.0? {1.0 + eps_computed == 1.0}")
print(f"   1.0 + 2ε  = {1.0 + 2*eps_computed:.17f}  → видно отличие ✓")

# === 4. Практическая иллюстрация: почему это важно ===
print(f"\n💡 Практический пример:")
a = 1.0
b = 1.0 + eps_computed/2  # Меньше ε — отличие теряется!
print(f"   a = {a:.17f}")
print(f"   b = 1.0 + ε/2 = {b:.17f}")
print(f"   a == b ? {a == b}  ← ⚠️ Компьютер 'не видит' разницы!")

📋 МАШИННЫЙ ЭПСИЛОН (ε) — минимальное число, для которого 1+ε > 1:

Тип данных      |      Машинный ε |   Значимых цифр
----------------------------------------------------
sys.float_info  |        2.22e-16 |             16 знаков
float64         |        2.22e-16 |             16 знаков
float32         |        1.19e-07 |              7 знаков
float16         |        9.77e-04 |              3 знаков

🔍 Вычисляем ε для float64 алгоритмически:
   Алгоритм: делим ε пополам, пока 1+ε не перестанет отличаться от 1
   Итераций: 52
   🏎️  Вычисленный ε = 2.22e-16

✅ Проверка результата:
   1.0 + ε/2 = 1.00000000000000000  → равно 1.0? True
   1.0 + ε   = 1.00000000000000022  → равно 1.0? False
   1.0 + 2ε  = 1.00000000000000044  → видно отличие ✓

💡 Практический пример:
   a = 1.00000000000000000
   b = 1.0 + ε/2 = 1.00000000000000000
   a == b ? True  ← ⚠️ Компьютер 'не видит' разницы!


### ✅ Ключевые выводы:

| Тип | Значимых цифр | Когда использовать |
|-----|--------------|-------------------|
| **float16** | 3 | Графика, ML-инференс (экономия памяти) |
| **float32** | 7 | Графика, ML-тренировка, GPU-вычисления |
| **float64** | 16 | 🔬 Научные расчёты, численные методы |

### 🎯 Практические правила:

1.  **Сравнение чисел:** `np.allclose(a, b)` вместо `a == b`

Внимание , ловушка !

In [ ]:
# === 🎯 Демонстрация: классическая ловушка плавающей запятой ===

# Исходные числа (в десятичной записи выглядят "простыми")
a = 0.1
b = 0.2
c = 0.3

print("=== 🔍 Проверка: 0.1 + 0.2 == 0.3 ? ===\n")

# Наивное сравнение — классическая ошибка новичков!
print(f"0.1 + 0.2 = {a + b}")
print(f"0.3       = {c}")
print(f"Равны ли? {a + b == c}")  # ❌ False!

# === Покажем "скрытые" цифры: формат с 17 знаками после запятой ===
# (17 знаков — чтобы увидеть ошибку округления float64)
print(f"\n=== 🔬 Смотрим глубже (17 знаков после запятой): ===")
print(f"0.1 + 0.2 = {format(a + b, '.17f')}")
print(f"0.3       = {format(c, '.17f')}")
print(f"Разница   = {format((a + b) - c, '.17e')}")

# === ✅ Как правильно сравнивать вещественные числа ===
print(f"\n=== ✅ Правильное сравнение: ===")

# Способ 1: Абсолютный допуск (если числа близки к нулю)
tol_abs = 1e-12
print(f"abs((a+b) - c) < {tol_abs:.0e} ? {abs((a + b) - c) < tol_abs}")

# Способ 2: Относительный допуск (универсальный)
tol_rel = 1e-9
relative_diff = abs((a + b) - c) / max(1.0, abs(a + b), abs(c))
print(f"Относит. разница = {relative_diff:.2e} < {tol_rel:.0e} ? {relative_diff < tol_rel}")

# Способ 3: Использовать готовую функцию NumPy (рекомендуется!)
import numpy as np
print(f"np.allclose(0.1+0.2, 0.3) ? {np.allclose(a + b, c)}")

=== 🔍 Проверка: 0.1 + 0.2 == 0.3 ? ===

0.1 + 0.2 = 0.30000000000000004
0.3       = 0.3
Равны ли? False

=== 🔬 Смотрим глубже (17 знаков после запятой): ===
0.1 + 0.2 = 0.30000000000000004
0.3       = 0.29999999999999999
Разница   = 5.55111512312578270e-17

=== ✅ Правильное сравнение: ===
abs((a+b) - c) < 1e-12 ? True
Относит. разница = 5.55e-17 < 1e-09 ? True
np.allclose(0.1+0.2, 0.3) ? True


### 🔑 Что мы увидели:

| Выражение | Значение (17 знаков) | Комментарий |
|-----------|---------------------|-------------|
| `0.1 + 0.2` | `0.30000000000000004` | Накопилась ошибка округления |
| `0.3`       | `0.29999999999999999` | Тоже приближённое значение |
| **Разница** | `~5.55·10⁻¹⁷` | Порядка машинного ε — но `==` этого не прощает! |

### 🎯 Как правильно работать с вещественными числами:

```python
# ❌ НИКОГДА не делайте так:
if a + b == c: ...

# ✅ Используйте допуски:
if abs((a + b) - c) < 1e-12: ...  # Абсолютный допуск

# ✅ Или относительный допуск (универсальнее):
if abs((a + b) - c) < 1e-9 * max(1.0, abs(a + b), abs(c)): ...

# ✅ Лучший вариант: готовая функция NumPy
import numpy as np
if np.allclose(a + b, c, rtol=1e-9, atol=1e-12): ...

Пример сложения больших и малых чисел

In [ ]:
# === 🎯 Демонстрация: потеря малого слагаемого ===
import numpy as np

# Числа разных порядков
big = 1e16      # Большое число
small = 1.0     # Малое число (на 16 порядков меньше!)



# Сценарий 1: big + small
s1 = big + small
print(f"big + small         = {s1:.1f}")
print(f"  (должно быть 10000000000000001.0)")

# Сценарий 2: big + small + small
s2 = big + small + small
print(f"big + small + small = {s2:.1f}")
print(f"  (должно быть 10000000000000002.0)")

# Проверяем разность: если small "потерялся", разность будет 0
print(f"\nРазность s2 - s1    = {s2 - s1:.1f}")
print(f"  (должно быть 1.0, но получили {s2 - s1})")

# === 🔬 Демонстрация "исчезновения" в явном виде ===
print(f"\n=== 🎯 Сдвигаемся в масштаб: (big + small) - big ===")
big2 = 1e16
small2 = 1.0
result = (big2 + small2) - big2

print(f"small2              = {small2:.1f}")
print(f"(big2 + small2) - big2 = {result:.1f}")
print(f"  ⚠️ Малое число 'исчезло' при сложении с большим!")


big + small         = 10000000000000000.0
  (должно быть 10000000000000001.0)
big + small + small = 10000000000000000.0
  (должно быть 10000000000000002.0)

Разность s2 - s1    = 0.0
  (должно быть 1.0, но получили 0.0)

=== 🎯 Сдвигаемся в масштаб: (big + small) - big ===
small2              = 1.0
(big2 + small2) - big2 = 0.0
  ⚠️ Малое число 'исчезло' при сложении с большим!


> 📌 **Золотое правило:**  
> При работе с числами разных порядков — масштабируйте задачу или используйте алгоритмы с компенсацией ошибок!


### Очень простой демонстрационный пример с «плохим» и «хорошим» порядком суммирования:

### 🔄 Порядок суммирования: почему он важен для точности?

**Проблема:** При последовательном сложении `s += x` малые слагаемые могут **теряться**, если текущая сумма `s` уже велика.

**Почему?**  
Если $|x| < |s| \cdot \varepsilon$, то $\text{fl}(s + x) = s$ — малое число не влияет на результат!

**Решение:** Складывать числа **от меньшего к большему** — так малые вклады успевают "накопиться" до значимой величины.

> 💡 **Эксперимент:** 10 чисел по $10^{16}$ + 1 000 000 чисел по $1.0$.  
> Теоретическая сумма: $10^{17} + 10^6 = 10000000000000000

In [ ]:
# === 🎯 Демонстрация: влияние порядка суммирования на точность ===
import numpy as np

# Исходные данные: 10 больших чисел + 1 миллион малых
values = [1e16] * 10 + [1.0] * 1_000_000

# Теоретическая точная сумма (в математике)
exact_sum = 10 * 1e16 + 1_000_000 * 1.0


# --- ❌ ПЛОХОЙ порядок: как есть (большие числа первыми) ---
# Малые слагаемые добавляются к уже огромной сумме → теряются из-за ε
s_bad = 0.0
for v in values:
    s_bad += v

# --- ✅ ХОРОШИЙ порядок: сначала малые, потом большие ---
# Малые числа складываются между собой → накапливаются → становятся значимыми
s_good = sum(sorted(values))  # sorted() по возрастанию: [1.0, 1.0, ..., 1e16]

# --- Сравнение результатов ---
print("=== 📊 Результаты суммирования ===")
print(f"Плохая сумма (исходный порядок): {s_bad:.1f}")
print(f"Хорошая сумма (по возрастанию) : {s_good:.1f}")
print(f"Теоретическая сумма            : {exact_sum:.1f}")

print(f"\n=== 🔍 Погрешности ===")
err_bad = abs(s_bad - exact_sum)
err_good = abs(s_good - exact_sum)
print(f"Ошибка плохого порядка : {err_bad:.2e}  ({err_bad/exact_sum*100:.2e}%)")
print(f"Ошибка хорошего порядка: {err_good:.2e}  ({err_good/exact_sum*100:.2e}%)")


# === 🔬 Бонус: почему сортировка помогает? ===
print(f"\n=== 💡 Механика потерь ===")
eps = np.finfo(float).eps
big = 1e16
print(f"Машинный ε              = {eps:.2e}")
print(f"Порог потери для {big:.1e}: |x| < {big*eps:.1f}")
print(f"Наше малое слагаемое    = 1.0")
print(f"1.0 < {big*eps:.1f} ? {1.0 < big*eps} → ⚠️ теряется при добавлении к {big:.1e}")

=== 📊 Результаты суммирования ===
Плохая сумма (исходный порядок): 100000000000000000.0
Хорошая сумма (по возрастанию) : 100000000001000000.0
Теоретическая сумма            : 100000000001000000.0

=== 🔍 Погрешности ===
Ошибка плохого порядка : 1.00e+06  (1.00e-09%)
Ошибка хорошего порядка: 0.00e+00  (0.00e+00%)

=== 💡 Механика потерь ===
Машинный ε              = 2.22e-16
Порог потери для 1.0e+16: |x| < 2.2
Наше малое слагаемое    = 1.0
1.0 < 2.2 ? True → ⚠️ теряется при добавлении к 1.0e+16


Проблема обычного суммирования: при сложении большого количества чисел ошибки округления накапливаются, особенно если числа сильно различаются по порядку.
​

Идея Кахана: ведётся учёт «потерянной» при округлении части суммы и добавляется она в следующий шаг. Это компенсирует ошибки округления


s ← 0     # основная сумма

c ← 0     # компенсация (потерянная часть)




In [ ]:
# === 🛡️ Алгоритм Кахана: компенсированное суммирование ===
import numpy as np

def kahan_sum(arr: np.ndarray) -> float:
    """
    Алгоритм Кахана для точного суммирования чисел с плавающей запятой.

    Принцип работы:
    ---------------
    На каждом шаге мы отслеживаем ошибку округления и компенсируем её
    при добавлении следующего слагаемого.

    Переменные:
    - s: текущая сумма (sum)
    - c: накопленная ошибка округления (compensation)
    - y: слагаемое с учётом компенсации
    - t: новая сумма после добавления
    """
    # Инициализируем в float64 для явной точности
    s = np.float64(0.0)  # Сумма
    c = np.float64(0.0)  # Компенсация (накопленная ошибка)

    for x in arr:
        # Шаг 1: Учитываем накопленную ошибку из предыдущего шага
        y = np.float64(x) - c  # "Очищенное" слагаемое

        # Шаг 2: Добавляем к сумме (здесь происходит округление!)
        t = s + y  # Новая сумма

        # Шаг 3: Вычисляем, сколько потеряли при округлении
        # (t - s) — это то, что фактически добавилось к s
        # y — это то, что мы хотели добавить
        # Разница — это потерянная часть
        c = (t - s) - y  # Новая ошибка для компенсации

        # Шаг 4: Обновляем сумму
        s = t

    return s


# === 🧪 Эксперимент: сравнение с обычным суммированием ===
# Создаём тестовые данные: 1 большое + 1 миллион малых
values = [1e16] * 1 + [0.5] * 1_000_000 +   [0.5] * 1_000_000

# Теоретически точная сумма (математика)
exact_sum = 1e16 + 1_000_000 * 1.0

# Сравнение трёх методов
print("=== 📊 Сравнение методов суммирования ===")
print(f"{'Метод':<25} | {'Результат':>22} | {'Ошибка':>12}")
print("-" * 65)


# 1. NumPy sum (использует частичную компенсацию)
sum_numpy = np.sum(values)
err_numpy = abs(sum_numpy - exact_sum)
print(f"{'np.sum()':<25} | {sum_numpy:22.2f} | {err_numpy:12.6f}")

# 2. Алгоритм Кахана (наша реализация)
sum_kahan = kahan_sum(np.array(values))
err_kahan = abs(sum_kahan - exact_sum)
print(f"{'Kahan sum ':<25} | {sum_kahan:22.2f} | {err_kahan:12.6f}")



=== 📊 Сравнение методов суммирования ===
Метод                     |              Результат |       Ошибка
-----------------------------------------------------------------
np.sum()                  |   10000000000999992.00 |     8.000000
Kahan sum                 |   10000000001000000.00 |     0.000000


Надо сказать. что в numpy  для float типа используется алгоритм частичного парного суммирования.

Если же нужно еще более стабилное вычисление суммы, то можно использовать math.fsum.  Python-функция math.fsum() представляет собой высокоточную методику суммирования чисел с плавающей точкой. Она использует специальный алгоритм ( Shewchuk в Python), который минимизирует потерю точности, характерную для стандартной арифметики с плавающей точкой.

## ⚡ Бенчмарк методов суммирования: скорость vs. точность на 100 млн чисел

**Цель:** Сравнить три подхода к суммированию на реалистичном объёме данных:

| Метод | Алгоритм | Точность | Скорость | Когда использовать |
|-------|----------|----------|----------|-------------------|
| **`sum()`** | Последовательное сложение | ⚠️ Средняя | 🟡 Средняя | Прототипы, небольшие данные |
| **`np.sum()`** | Векторизация + SIMD + многопоточность | ⚠️ Зависит от порядка | 🚀 Быстро | Большие массивы, однородные данные |
| **`math.fsum()`** | Алгоритм Шьючука (точное округление) | 🏆 Максимальная | 🐌 Медленно | Финансы, научные расчёты, где важна каждая цифра |

**Тестовый сценарий:**  
$3 \times 1.0 + 100\,000\,000 \times 10^{-8} = 3 + 1 = 4$  
Малые слагаемые в $10^8$ раз меньше больших — идеальная проверка на потерю значимости!


In [ ]:
# === ⚡ Бенчмарк: сравнение методов суммирования ===
import numpy as np
import math
import time

def benchmark_sums(n_big=3, n_small=100_000_000):
    """
    Сравнивает точность и скорость трёх методов суммирования.

    Параметры:
    ----------
    n_big : int
        Количество больших чисел (1.0)
    n_small : int
        Количество малых чисел (1e-8)
    """
    print(f"\n{'='*70}")
    print(f"🚨 BENCHMARK: {n_big}×1.0 + {n_small:,}×1e-8")
    print(f"{'='*70}")

    # === 1. Подготовка данных ===
    # Большие числа идут первыми — худший порядок для точности!
    numbers_desc = np.full(n_big, 1.0, dtype=np.float64)
    numbers_asc = np.full(n_small, 1e-8, dtype=np.float64)

    # Конкатенируем: сначала большие, потом малые (провоцируем потерю точности)
    numbers = np.concatenate([numbers_desc, numbers_asc])

    # Теоретически точная сумма (математика, без округлений)
    s_theory = n_big * 1.0 + n_small * 1e-8
    print(f"🎯 Теоретическая сумма: {s_theory:.10f}\n")

    # === 2. Словарь методов для тестирования ===
    methods = {
        'sum()': lambda x: sum(x),           # Python built-in
        'np.sum()': lambda x: np.sum(x),     # NumPy векторизованный
        'math.fsum()': lambda x: math.fsum(x) # Точное суммирование
    }

    results = {}
    times = {}

    # === 3. Замер времени и точности для каждого метода ===
    print(f"{'Метод':<12} | {'Результат':>18} | {'Ошибка Δ':>12} | {'Время':>8} | {'Статус'}")
    print("-" * 70)

    for name, func in methods.items():
        start = time.perf_counter()
        result = func(numbers)
        end = time.perf_counter()

        results[name] = result
        times[name] = (end - start) * 1000  # Время в миллисекундах
        delta = abs(result - s_theory)      # Абсолютная ошибка

        # Статус: ✅ если ошибка близка к машинному эпсилону
        status = "✅" if delta < 1e-10 else "❌"
        print(f"{name:<12} | {result:18.15f} | {delta:12.2e} | {times[name]:6.2f}мс | {status}")

    # === 4. Итоговая сводка ===
    print(f"\n{'='*70}")
    print(f"📊 ИТОГИ:")

    # Находим лучший по точности и по скорости
    best_accuracy = min(results.keys(), key=lambda k: abs(results[k] - s_theory))
    best_speed = min(times.keys(), key=lambda k: times[k])

    print(f"• 🎯 Наилучшая точность: {best_accuracy}")
    print(f"• 🚀 Наивысшая скорость:  {best_speed}")
    print(f"{'='*70}\n")

    return results, times

# === 🚀 Запуск бенчмарка ===
# Внимание: n_small=100_000_000 может занять ~10-30 секунд на math.fsum()
benchmark_sums(n_big=3, n_small=100_000_000)


🚨 BENCHMARK: 3×1.0 + 100,000,000×1e-8
🎯 Теоретическая сумма: 4.0000000000

Метод        |          Результат |     Ошибка Δ |    Время | Статус
----------------------------------------------------------------------
sum()        |  3.999999993922529 |     6.08e-09 | 10987.77мс | ❌
np.sum()     |  3.999999999999480 |     5.20e-13 |  75.28мс | ✅
math.fsum()  |  4.000000000000000 |     0.00e+00 | 8268.75мс | ✅

📊 ИТОГИ:
• 🎯 Наилучшая точность: math.fsum()
• 🚀 Наивысшая скорость:  np.sum()



({'sum()': np.float64(3.999999993922529),
  'np.sum()': np.float64(3.9999999999994795),
  'math.fsum()': 4.0},
 {'sum()': 10987.765399999944,
  'np.sum()': 75.27967099997568,
  'math.fsum()': 8268.749954999976})


### 📈 Анализ результатов:

| Метод | Ошибка | Время | Почему так? |
|-------|--------|-------|-------------|
| **`sum()`** | ~$6\cdot10^{-8}$ | ~30 мс | Детерминированный порядок, но нет компенсации ошибок |
| **`np.sum()`** | ~$5\cdot10^{-13}$ | ~8 мс | ⚡ Векторизация + SIMD, но порядок операций недетерминирован → ошибка тоже присутствует |
| **`math.fsum()`** | **0** | ~850 мс | 🏆 Алгоритм Шьючука: точное округление на каждом шаге, но в 25-100× медленнее |

### 🎯 Практические рекомендации:

```python
# 🔹 Нужна скорость и данные однородные?
→ np.sum(arr)  # Быстро, ошибка приемлема

# 🔹 Нужна гарантия точности (финансы, наука)?
→ math.fsum(arr)  # Медленно, но точно

# 🔹 Числа различаются на >10⁶ порядков?
→ Kahan sum или math.fsum, иначе потеря значимости неизбежна

## Разбор структуры числа в формате IEEE 754 double precision

In [ ]:
import struct
import numpy as np

pi = np.pi
# Биты double
bits = struct.unpack('Q', struct.pack('d', pi))[0]

print(f"π в double (bin): {bits:064b}")

# Разбор по полям
sign = (bits >> 63) & 1
exponent = (bits >> 52) & 0x7FF
mantissa = bits & 0xFFFFFFFFFFFFF

print(f"\nЗНАК:     {sign} ({'+' if sign==0 else '-'})")
print(f"ПОРЯДОК:  {exponent} (истинный = {exponent-1023})")
print(f"МАНТИССА: {mantissa:064b} (скрытая 1.)")
print(f"МАНТИССА: {mantissa} (скрытая 1.)")


π в double (bin): 0100000000001001001000011111101101010100010001000010110100011000

ЗНАК:     0 (+)
ПОРЯДОК:  1024 (истинный = 1)
МАНТИССА: 0000000000001001001000011111101101010100010001000010110100011000 (скрытая 1.)
МАНТИССА: 2570638124657944 (скрытая 1.)




Получаем:

ЗНАК = 0  
$(-1)^{0}$ = +1 ✓  

ПОРЯДОК = 1024  
$1024 - 1023 = 1 \rightarrow 2^{1}$ ✓  

МАНТИССА  

Перевод мантиссы в десятичные дроби (первые несколько бит):  

$$
1.0000000000001001001000011111101101010100010001000010110100011000_2 \\
= 1 + 2^{-13} + 2^{-16} + 2^{-18} + 2^{-20} + 2^{-21} + \dotsb \\
\approx 1 + 0.0001220703125 + 0.0000152587890625 + \dotsb \\
\approx 1.5707963267948966_{10} \quad✓
$$

Полное число:  

$$
\pi \approx (+1) \times 1.5707963267948966_{10} \times 2^{1} \\
= 1.5707963267948966_{10} \times 2 \\
= 3.141592653589793_{10}\quad ✓
$$

In [ ]:
# Собираем обратно
sign, exp, mant = 0, 1024, 2570638124657944
true_exp = exp - 1023  # = 1

# Мантисса как float: 1 + mantissa/(2^52)
mantissa_float = 1.0 + mant / (2**52)
pi_reconstructed = ((-1)**sign) * (2**true_exp) * mantissa_float

print(f"Восстановленное π: {pi_reconstructed}")
print(f"np.pi:              {np.pi}")
print(f"Равны:              {abs(pi_reconstructed - np.pi) < 1e-15}")


Восстановленное π: 3.141592653589793
np.pi:              3.141592653589793
Равны:              True


##  Умножение чисел

## 🧮 Устойчивое умножение: как избежать переполнения и исчезновения порядка?

**Проблема:** При перемножении многих чисел промежуточный результат может:
- **Переполниться** (→ `inf`), если множим много чисел > 1
- **Исчезнуть** (→ `0`), если множим много чисел < 1



In [ ]:
# === 🧮 Устойчивые алгоритмы умножения ===
import numpy as np
import math

def regular_product(nums):
    """Наивное умножение — быстро, но может переполниться"""
    res = 1.0
    for n in nums:
        res *= n
    return res

def log_product(nums, return_log=False):
    """
    Логарифмическое умножение.

    Параметры:
    ----------
    return_log : bool
        Если True, возвращаем log(произведения) — это всегда безопасно!
    """
    nums = np.array(nums, dtype=np.float64)
    if np.any(nums == 0):
        return 0.0 if not return_log else -np.inf

    sign = np.prod(np.sign(nums))
    log_sum = np.sum(np.log(np.abs(nums)))

    # Если просят логарифм — возвращаем сразу (без exp!)
    if return_log:
        return log_sum if sign > 0 else np.nan

    # Проверка: exp() переполнится при log_sum > 709
    if log_sum > 709:
        return np.inf
    elif log_sum < -745:
        return 0.0

    return sign * np.exp(log_sum)


# === 🧪 Пример 1: Переполнение (regular fails, log в логах работает) ===
print("="*60)
print("🧪 Пример 1: Переполнение (10^310 > 10^308)")
print("="*60)

test1 = [10.0] * 310  # 10^310 — больше предела float64 (~10^308)

reg = regular_product(test1)
log_val = log_product(test1)
log_sum = log_product(test1, return_log=True)  # ✅ Безопасно!

print(f"regular_product: {reg}  ❌ overflow (inf)")
print(f"log_product:     {log_val}  ⚠️ тоже inf (exp(713) > 709)")
print(f"\n✅ РЕШЕНИЕ: остаться в лог-пространстве!")
print(f"   log(произведения) = {log_sum:.2f}")
print(f"   Это корректно! (10^310 → log = 310·log(10) ≈ 713)")


# === 🧪 Пример 2: Исчезновение (regular fails, log works) ===
print(f"\n{'='*60}")
print("🧪 Пример 2: Исчезновение (0.01^200 < 10^-308)")
print("="*60)

test2 = [0.01] * 200  # 0.01^200 = 10^-400 — меньше предела float64

reg = regular_product(test2)
log_val = log_product(test2)

print(f"regular_product: {reg}  ❌ underflow (0.0)")
print(f"log_product:     {log_val:.4e}  ✅ корректно!")
print(f"Теоретически:    0.01^200 = 10^-400")


# === 🧪 Пример 3: Накопление ошибки (log точнее) ===
print(f"\n{'='*60}")
print("🧪 Пример 3: Накопление ошибки (10 000 умножений)")
print("="*60)

test3 = [1.001] * 10000  # 1.001^10000 ≈ 22026

reg = regular_product(test3)
log_val = log_product(test3)
theory = 1.001**10000

err_reg = abs(reg - theory) / theory
err_log = abs(log_val - theory) / theory

print(f"regular_product: {reg:.4f}  (ошибка: {err_reg:.2e})")
print(f"log_product:     {log_val:.4f}  (ошибка: {err_log:.2e})")
print(f"✅ log_product точнее в {err_reg/err_log:.0f} раз!")


# === 🧪 Пример 4: ML-сценарий (произведение вероятностей) ===
print(f"\n{'='*60}")
print("🧪 Пример 4: ML — произведение 1000 вероятностей")
print("="*60)

probs = [0.5] * 1000  # 0.5^1000 ≈ 10^-301

reg = regular_product(probs)
log_val = log_product(probs)
log_likelihood = np.sum(np.log(probs))  # ✅ В ML используют это!

print(f"regular_product:   {reg:.4e}  ⚠️ риск underflow")
print(f"log_product:       {log_val:.4e}  ✅ надёжно")
print(f"\n🎯 В ML работают в логах:")
print(f"   log-likelihood = {log_likelihood:.2f}")
print(f"   (никогда не вычисляют произведение явно!)")



🧪 Пример 1: Переполнение (10^310 > 10^308)
regular_product: inf  ❌ overflow (inf)
log_product:     inf  ⚠️ тоже inf (exp(713) > 709)

✅ РЕШЕНИЕ: остаться в лог-пространстве!
   log(произведения) = 713.80
   Это корректно! (10^310 → log = 310·log(10) ≈ 713)

🧪 Пример 2: Исчезновение (0.01^200 < 10^-308)
regular_product: 0.0  ❌ underflow (0.0)
log_product:     0.0000e+00  ✅ корректно!
Теоретически:    0.01^200 = 10^-400

🧪 Пример 3: Накопление ошибки (10 000 умножений)
regular_product: 21916.6813  (ошибка: 6.47e-15)
log_product:     21916.6813  (ошибка: 3.32e-15)
✅ log_product точнее в 2 раз!

🧪 Пример 4: ML — произведение 1000 вероятностей
regular_product:   9.3326e-302  ⚠️ риск underflow
log_product:       9.3326e-302  ✅ надёжно

🎯 В ML работают в логах:
   log-likelihood = -693.15
   (никогда не вычисляют произведение явно!)


In [ ]:
nums = [0.99] * 200 + [1.01] * 50 + [1e-8, 1e8]

print("Наивное:      ", regular_product(nums))
print("np.prod:      ", np.prod(nums))

print("Логарифмы:    ", log_product(nums))


Наивное:       0.2203472367517007
np.prod:       0.2203472367517007
Логарифмы:     0.22034723675170045


## 🔬 Эталонное вычисление с произвольной точностью (mpmath)

**Зачем это нужно?**  
Когда мы тестируем устойчивые алгоритмы, нам нужно знать **точное значение**, с которым сравнивать результаты float64.

**mpmath:** Библиотека для арифметики произвольной точности (software floating-point).

| Тип | Точность | Скорость | Когда использовать |
|-----|----------|----------|-------------------|
| **float64** | ~16 знаков | 🚀 Быстро | Основные вычисления |
| **mpmath (50 dps)** | 50 знаков | 🐌 Медленно (×100-1000) | ✅ Эталон для тестов, отладка |
| **mpmath (100+ dps)** | 100+ знаков | 🐢 Очень медленно | Критические научные расчёты |

> 💡 **Важно:** mpmath — это программная эмуляция, а не аппаратная арифметика. Используйте только для валидации!

In [ ]:
# === 🔬 Эталонное вычисление через mpmath (произвольная точность) ===
from mpmath import mp

# Настраиваем точность: 50 десятичных знаков (digits of precision)
mp.dps = 50

print(f"🎯 Вычисляем произведение с точностью {mp.dps} знаков:\n")

# Исходные данные (пример: 1000 чисел 1.001)
nums = [1.001] * 1000

# --- 1. Вычисление в mpmath ---
# Важно: конвертируем через str(), чтобы избежать ошибок float64 при передаче!
prod_mp = mp.mpf(1)
for num in nums:
    prod_mp *= mp.mpf(str(num))  # str(num) → точная десятичная запись

# --- 2. Сравнение с float64 методами ---
reg = regular_product(nums)
log_val = log_product(nums)

# --- 3. Вывод результатов ---
print("=== 📊 Сравнение с эталоном (mpmath) ===")
print(f"mpmath (50 знаков):  {prod_mp}")  # mpmath сам форматирует при print
print(f"\nfloat64 результаты:")
print(f"  regular_product:   {reg:.15e}  (ошибка: {abs(reg - float(prod_mp)):.2e})")
print(f"  log_product:       {log_val:.15e}  (ошибка: {abs(log_val - float(prod_mp)):.2e})")

# --- 4. Бонус: показываем, где float64 "отрезает" знаки ---
print(f"\n=== 🔍 Детали: сколько знаков совпадает? ===")

# ИСПРАВЛЕНО: используем mp.nstr() для форматирования mpmath чисел
mp_str = mp.nstr(prod_mp, 20)  # 20 значащих цифр
float_str = f"{float(prod_mp):.20g}"

print(f"mpmath : {mp_str}")
print(f"float64: {float_str}")

# Сравниваем первые 16 символов (предел float64)
mp_clean = str(mp_str).replace('.', '')[:16]
float_clean = float_str.replace('.', '')[:16]
print(f"         {'✓' if mp_clean == float_clean else '✗'} Совпадают первые ~16 знаков (предел float64)")

🎯 Вычисляем произведение с точностью 50 знаков:

=== 📊 Сравнение с эталоном (mpmath) ===
mpmath (50 знаков):  2.7169239322358924573830881219475771889643150188376

float64 результаты:
  regular_product:   2.716923932235598e+00  (ошибка: 2.94e-13)
  log_product:       2.716923932235593e+00  (ошибка: 3.00e-13)

=== 🔍 Детали: сколько знаков совпадает? ===
mpmath : 2.7169239322358924574
float64: 2.7169239322358924582
         ✓ Совпадают первые ~16 знаков (предел float64)


## Суммирование большого числа чисел

In [ ]:
# === 🔄 Тест: влияние порядка суммирования на точность ===
import numpy as np

def test_sum_order(dtype=np.float32):
    """
    Сравнивает точность суммирования ряда 1/k² в разном порядке.

    Параметры:
    ----------
    dtype : np.dtype
        Тип данных для вычислений (float32 или float64)
    """
    print(f"\n{'═'*70}")
    print(f"🔥 ТЕСТ: ПОРЯДОК СУММИРОВАНИЯ ({dtype.__name__.upper()})")
    print(f"{'═'*70}")

    # Тестируем на разных масштабах: n = 10⁴, 10⁵, 10⁶
    for n_exp in [4, 5, 6]:
        n = 10**n_exp

        # ✅ ХОРОШИЙ порядок: малые слагаемые первыми
        # Генерируем: 1/n², 1/(n-1)², ..., 1/1² (возрастание)
        x_good = 1.0 / ((np.arange(n)[::-1] + 1)**2).astype(dtype)

        # ❌ ПЛОХОЙ порядок: большие слагаемые первыми
        # Генерируем: 1/1², 1/2², ..., 1/n² (убывание)
        x_bad = 1.0 / ((np.arange(n) + 1)**2).astype(dtype)

        # "Эталон": считаем в float64 для сравнения
        s_exact = np.sum(x_good.astype(np.float64))

        # Наивное суммирование Python (слева направо)
        s_naive_good = sum(x_good)  # малые → большие
        s_naive_bad = sum(x_bad)    # большие → малые

        # Вывод результатов
        print(f"\n📊 n = {n:,} | Точное S ≈ {s_exact:.6f}")
        print(f"✅ Малые→Большие: {s_naive_good:.8f}  | Ошибка Δ = {abs(s_naive_good - s_exact):.2e}")
        print(f"❌ Большие→Малые: {s_naive_bad:.8f}  | Ошибка Δ = {abs(s_naive_bad - s_exact):.2e}")



# Запуск теста для float32 (наиболее наглядный случай)
test_sum_order(np.float32)


══════════════════════════════════════════════════════════════════════
🔥 ТЕСТ: ПОРЯДОК СУММИРОВАНИЯ (FLOAT32)
══════════════════════════════════════════════════════════════════════

📊 n = 10,000 | Точное S ≈ 1.644834
✅ Малые→Большие: 1.64483404  | Ошибка Δ = 2.97e-08
❌ Большие→Малые: 1.64472532  | Ошибка Δ = 1.09e-04

📊 n = 100,000 | Точное S ≈ 1.644924
✅ Малые→Большие: 1.64492404  | Ошибка Δ = 2.17e-08
❌ Большие→Малые: 1.64472532  | Ошибка Δ = 1.99e-04

📊 n = 1,000,000 | Точное S ≈ 1.644933
✅ Малые→Большие: 1.64493299  | Ошибка Δ = 8.10e-08
❌ Большие→Малые: 1.64472532  | Ошибка Δ = 2.08e-04


## Округление

In [ ]:
## банковское округление
np.round([0.15, 0.25,0.35 , 0.45],1)

array([0.2, 0.2, 0.4, 0.4])

In [ ]:
import math

def fast_math_round(x, decimals=0):
    """Быстрое математическое округление"""
    multiplier = 10**decimals
    return math.copysign(1, x) * math.floor(abs(x) * multiplier + 0.5) / multiplier

print(fast_math_round(2.3456, 2))  # 2.35
print(fast_math_round(2.5, 0))     # 3.0
print(fast_math_round(-2.325, 2))  # -2.33


2.35
3.0
-2.33
